### PR TITLE
LUCENE-10375: Write vectors to file in flush

### PR DESCRIPTION
**Pull request by jtibshirani**
In a previous commit, we updated HNSW merge to first write the combined segment
vectors to a file, then use that file to build the graph. This commit applies
the same strategy to flush, which lets us use the same logic for flush and
merge.

### CODE REVIEW

**Comment by jtibshirani** [comment]
Some notes:
* The HNSW format no longer needs the ability to override `merge`. It's still possible though -- it seems like helpful flexibility and there's precedent with other formats.
* We still can't remove `RandomAccessVectorValues` or make it package-private to HNSW. This is because `SortingVectorValues` assumes its input supports random access. We can improve this later as part of [LUCENE-9583](https://issues.apache.org/jira/browse/LUCENE-9583).

**Comment by msokolov** [comment]
Well, it's certainly a nice cleanup. The performance results are confusing at first glance, since we now write the temp file and then copy it, but consider that now when we traverse the vectors while writing the graph, we are accessing a memory-mapped file that densely stores the vectors, where before we were accessing those same vectors through an array of pointers, requiring an extra lookup, so it seems that it's a net cost savings! awesome!

**Comment by jtibshirani** [comment]
Ah right, that makes sense. Somehow I thought there'd be significant overhead from decoding vectors from the on-disk format, but I guess that's not true.

Anyways, thanks for taking a look. I plan to merge in the next day if there aren't more comments.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90HnswVectorsWriter.java [hunk 1]
```diff
@@ -114,79 +113,15 @@
   public void writeField(FieldInfo fieldInfo, KnnVectorsReader knnVectorsReader)
       throws IOException {
     long vectorDataOffset = vectorData.alignFilePointer(Float.BYTES);
-
     VectorValues vectors = knnVectorsReader.getVectorValues(fieldInfo.name);
-    // TODO - use a better data structure; a bitset? DocsWithFieldSet is p.p. in o.a.l.index
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90HnswVectorsWriter.java:119]: nit: can you retain that TODO?

**Review by jpountz — APPROVED** [review]
Fantastic. I have the same intuition as @msokolov that benefits from improved data locality might exceed the decoding overhead.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90HnswVectorsWriter.java (hunk 1 above)
Comment (jtibshirani) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90HnswVectorsWriter.java:119]: Oops, will do.

**Pull request merged**